File: tests/io/select_eintr.js
// A listener makes Bun install a handler for SIGALRM (ignored by
// default, so a run that never got the handler is not killed); an
// interval timer has the kernel send it ms later, whatever the loop is
// doing then. A child process cannot promise that: a kill spawned here
// lands in about half a millisecond on Linux, most often before the
// loop reaches select, and the test then exercises nothing. Then the
// computation parks on the read end of a pipe whose write end stays
// open and unwritten: its fd never becomes ready, so a wake is
// spurious.
function idle_park(ms, k) {
  const ffi = require("bun:ffi");
  const sys = io_sys();
  const lib = ffi.dlopen(sys.mac ? "libSystem.dylib" : "libc.so.6", {
    pipe: { args: ["ptr"], returns: "i32" },
    setitimer: { args: ["i32", "ptr", "ptr"], returns: "i32" },
  }).symbols;
  process.on("SIGALRM", () => {});
  const it = new BigInt64Array([0n, 0n,
    BigInt(ms / 1000 | 0), BigInt(ms % 1000 * 1000)]);
  lib.setitimer(0, sys.ptr(it), null);
  const p = new Int32Array(2);
  lib.pipe(sys.ptr(p));
  sys.fcntl(p[0], 4, sys.fcntl(p[0], 3, 0) | (sys.mac ? 4 : 0x800));
  io_park_on(p[0], false, k, () => ({ $: CID(Unit) }));
  return undefined;
}

io_eff(CID(Idle.park), idle_park);

File: tests/io/select_eintr.c
#include <fcntl.h>

// A no-op handler for SIGWINCH, installed without SA_RESTART, and a
// thread that sends it to the loop's thread ms later. Then the
// computation parks on the read end of a pipe whose write end stays open
// and unwritten: its fd never becomes ready, so a wake is spurious.
static pthread_t idle_loop;

static void idle_hush(int sig) {
}

static void* idle_ring(void* ms) {
  struct timespec t = { (uintptr_t)ms / 1000, (uintptr_t)ms % 1000 * 1000000 };
  nanosleep(&t, NULL);
  pthread_kill(idle_loop, SIGWINCH);
  return NULL;
}

static Term idle_park_more(Env e, IoWork* w) {
  return term_pak(CID(Unit), 0);
}

Term idle_park_run(Env e, Term* f, IoWork* w) {
  struct sigaction sa = { .sa_handler = idle_hush };
  sigaction(SIGWINCH, &sa, NULL);
  idle_loop = pthread_self();
  pthread_t t;
  pthread_create(&t, NULL, idle_ring, (void*)(uintptr_t)(u32)f[0]);
  pthread_detach(t);
  int p[2];
  pipe(p);
  fcntl(p[0], F_SETFL, O_NONBLOCK);
  return io_wait_on(w, p[0], POLLIN, 0, idle_park_more);
}

static void __attribute__((constructor)) idle_park_use(void) {
  io_eff(CID(Idle.park), idle_park_run, 0);
}
